File: pkg/node/net.go
package node

import (
	"encoding/base64"
	"net"
	"strconv"
	"sync"
	"sync/atomic"

	"github.com/tamnd/bento/pkg/engine"
)

// netBridgeState backs node:net. It owns the Go listeners and connections and
// maps them to the integer ids the JavaScript side holds, the same callback-by-id
// protocol http uses. A connection carries its own ordered write queue so writes
// from JavaScript reach the socket in order without ever blocking the loop.
type netBridgeState struct {
	netBridge
	mu      sync.Mutex
	nextID  atomic.Int64
	servers map[int64]net.Listener
	conns   map[int64]*netConn
}

// netConn is one live connection. Writes are handed to a single writer goroutine
// through writes so they serialize; end closes writes after draining, and destroy
// tears the socket down immediately.
type netConn struct {
	id     int64
	conn   net.Conn
	writes chan []byte
	once   sync.Once
}

func installNet(eng engine.Engine, loop LoopHost) error {
	n := &netBridgeState{
		netBridge: netBridge{eng: eng, loop: loop},
		servers:   map[int64]net.Listener{},
		conns:     map[int64]*netConn{},
	}
	for name, fn := range n.hostFuncs() {
		if err := eng.Register(name, fn); err != nil {
			return err
		}
	}
	return nil
}

func (n *netBridgeState) hostFuncs() map[string]HostFunc {
	return map[string]HostFunc{
		"__bento_net_createServer": n.createServer,
		"__bento_net_listen":       n.listen,
		"__bento_net_closeServer":  n.closeServer,
		"__bento_net_connect":      n.connect,
		"__bento_net_write":        n.write,
		"__bento_net_end":          n.end,
		"__bento_net_destroy":      n.destroy,
	}
}

func (n *netBridgeState) createServer(_ []any) (any, error) {
	id := n.nextID.Add(1)
	n.mu.Lock()
	n.servers[id] = nil
	n.mu.Unlock()
	return id, nil
}

// listen binds a TCP listener and accepts on a pool goroutine. Each accepted
// connection is registered and announced with a connection event, then its read
// and write pumps start. Binding is inline so a bind error surfaces before the
// loop commits to the server.
func (n *netBridgeState) listen(args []any) (any, error) {
	id := int64(intArg(args, 0))
	port := intArg(args, 1)
	host := str(args, 2)

	ln, err := net.Listen("tcp", net.JoinHostPort(host, strconv.Itoa(port)))
	if err != nil {
		n.emit("__bento_net_dispatchServerError", id, errCode(err), err.Error())
		return nil, nil
	}
	n.mu.Lock()
	n.servers[id] = ln
	n.mu.Unlock()

	bound := ln.Addr().(*net.TCPAddr)
	n.loop.AddRef()
	n.emit("__bento_net_dispatchListening", id, int64(bound.Port), bound.IP.String())

	n.pool(func() {
		for {
			conn, aerr := ln.Accept()
			if aerr != nil {
				break
			}
			nc := n.register(conn)
			n.emit("__bento_net_dispatchConnection", id, nc.id, connInfo(conn))
			n.startPumps(nc)
		}
		n.mu.Lock()
		delete(n.servers, id)
		n.mu.Unlock()
		n.loop.Post(func() { n.loop.Unref() })
		n.emit("__bento_net_dispatchServerClose", id)
	})
	return nil, nil
}

func (n *netBridgeState) closeServer(args []any) (any, error) {
	id := int64(intArg(args, 0))
	n.mu.Lock()
	ln := n.servers[id]
	n.mu.Unlock()
	if ln != nil {
		_ = ln.Close()
	}
	return nil, nil
}

// connect dials an outbound TCP connection. JavaScript mints the connection id up
// front so the connect and later data events can be routed. Success emits connect
// and starts the pumps; failure emits an error.
func (n *netBridgeState) connect(args []any) (any, error) {
	id := int64(intArg(args, 0))
	port := intArg(args, 1)
	host := str(args, 2)

	// Hold the loop open for the dialing window. This ref runs on the loop
	// goroutine here; on success the connection reuses it, on failure it drops.
	n.loop.AddRef()
	n.pool(func() {
		conn, err := net.Dial("tcp", net.JoinHostPort(host, strconv.Itoa(port)))
		if err != nil {
			n.loop.Post(func() { n.loop.Unref() })
			n.emit("__bento_net_dispatchError", id, err.Error())
			return
		}
		nc := n.adopt(id, conn, false)
		n.emit("__bento_net_dispatchConnect", id, connInfo(conn))
		n.startPumps(nc)
	})
	return nil, nil
}

func (n *netBridgeState) write(args []any) (any, error) {
	nc := n.lookup(int64(intArg(args, 0)))
	if nc == nil {
		return nil, nil
	}
	if data, err := base64.StdEncoding.DecodeString(str(args, 1)); err == nil {
		select {
		case nc.writes <- data:
		default:
			// The writer goroutine has exited (connection gone); drop the chunk.
		}
	}
	return nil, nil
}

// end closes the write half after queued writes drain, signalled by closing the
// writes channel. The read side stays open so a peer can still reply.
func (n *netBridgeState) end(args []any) (any, error) {
	nc := n.lookup(int64(intArg(args, 0)))
	if nc == nil {
		return nil, nil
	}
	nc.once.Do(func() { close(nc.writes) })
	return nil, nil
}

func (n *netBridgeState) destroy(args []any) (any, error) {
	nc := n.lookup(int64(intArg(args, 0)))
	if nc == nil {
		return nil, nil
	}
	_ = nc.conn.Close()
	return nil, nil
}

// register allocates an id for an inbound connection and records it, taking a
// fresh loop reference for the connection's lifetime.
func (n *netBridgeState) register(conn net.Conn) *netConn {
	return n.adopt(n.nextID.Add(1), conn, true)
}

// adopt records a connection under a given id. When addRef is set it posts a loop
// reference for the connection (inbound path); the connect path passes false
// because it already holds the dialing reference the connection reuses. AddRef
// is posted because adopt may run on an accept pool goroutine, off the loop.
func (n *netBridgeState) adopt(id int64, conn net.Conn, addRef bool) *netConn {
	nc := &netConn{id: id, conn: conn, writes: make(chan []byte, 64)}
	n.mu.Lock()
	n.conns[id] = nc
	n.mu.Unlock()
	if addRef {
		n.loop.Post(func() { n.loop.AddRef() })
	}
	return nc
}

func (n *netBridgeState) lookup(id int64) *netConn {
	n.mu.Lock()
	defer n.mu.Unlock()
	return n.conns[id]
}

// startPumps runs the read and write loops for a connection on pool goroutines.
// The read loop turns socket bytes into data and end events; the write loop
// serializes queued writes and half-closes on drain. When the read loop ends the
// connection is torn down and its loop reference dropped.
func (n *netBridgeState) startPumps(nc *netConn) {
	n.pool(func() {
		for data := range nc.writes {
			if _, err := nc.conn.Write(data); err != nil {
				break
			}
		}
		if tcp, ok := nc.conn.(*net.TCPConn); ok {
			_ = tcp.CloseWrite()
		}
	})

	n.pool(func() {
		buf := make([]byte, 64*1024)
		for {
			read, err := nc.conn.Read(buf)
			if read > 0 {
				n.emit("__bento_net_dispatchData", nc.id, base64.StdEncoding.EncodeToString(buf[:read]))
			}
			if err != nil {
				break
			}
		}
		n.emit("__bento_net_dispatchEnd", nc.id)
		_ = nc.conn.Close()
		n.mu.Lock()
		delete(n.conns, nc.id)
		n.mu.Unlock()
		n.loop.Post(func() { n.loop.Unref() })
		n.emit("__bento_net_dispatchClose", nc.id)
	})
}

// connInfo is the address snapshot handed to JavaScript for a socket, marshaled
// as JSON so the net module can populate remoteAddress and friends.
func connInfo(conn net.Conn) string {
	type info struct {
		RemoteAddress string `json:"remoteAddress"`
		RemotePort    int    `json:"remotePort"`
		LocalAddress  string `json:"localAddress"`
		LocalPort     int    `json:"localPort"`
	}
	var out info
	if a, ok := conn.RemoteAddr().(*net.TCPAddr); ok {
		out.RemoteAddress, out.RemotePort = a.IP.String(), a.Port
	}
	if a, ok := conn.LocalAddr().(*net.TCPAddr); ok {
		out.LocalAddress, out.LocalPort = a.IP.String(), a.Port
	}
	return jsonString(out)
}

File: pkg/node/js/net.js
// net implements the node:net TCP surface: Server and Socket over the Go
// networking bridge (pkg/node/net.go). Socket is a Duplex stream, so data flows
// through the same read and write machinery the rest of the stream module uses.
// The Go side owns the real listener and connections and drives this module
// through the __bento_net_dispatch* globals; this side drives writes back through
// the __bento_net_* host functions, all keyed by integer ids.

__bento_defineModule("net", function (module, exports, require) {
  "use strict";

  const EventEmitter = require("events");
  const stream = require("stream");

  const servers = Object.create(null); // serverId -> Server
  const sockets = Object.create(null); // connId -> Socket

  // Socket is a Duplex over one connection id. Reads arrive as pushed data and
  // writes are queued to Go in order. Both inbound (server) and outbound (connect)
  // connections use it; an outbound socket mints its own id before dialing.
  class Socket extends stream.Duplex {
    constructor(connId) {
      super();
      this._connId = connId;
      this._connecting = false;
      this.remoteAddress = undefined;
      this.remotePort = undefined;
      this.localAddress = undefined;
      this.localPort = undefined;
      this.bytesRead = 0;
      this.bytesWritten = 0;
    }
    _applyInfo(info) {
      this.remoteAddress = info.remoteAddress;
      this.remotePort = info.remotePort;
      this.localAddress = info.localAddress;
      this.localPort = info.localPort;
    }
    _read() {}
    _write(chunk, encoding, callback) {
      const buf = Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk, encoding || "utf8");
      this.bytesWritten += buf.length;
      __bento_net_write(this._connId, buf.toString("base64"));
      callback();
    }
    _final(callback) {
      __bento_net_end(this._connId);
      callback();
    }
    destroy(err) {
      __bento_net_destroy(this._connId);
      if (err) this.emit("error", err);
      return this;
    }
    setNoDelay() {
      return this;
    }
    setKeepAlive() {
      return this;
    }
    setTimeout(_msecs, callback) {
      if (callback) this.on("timeout", callback);
      return this;
    }
    address() {
      return { address: this.localAddress, port: this.localPort, family: "IPv4" };
    }
    ref() {
      return this;
    }
    unref() {
      return this;
    }
  }

  // connect accepts every net.connect shape: connect(port[, host][, cb]),
  // connect(options[, cb]). The trailing function, if any, is the connect
  // listener.
  function connect() {
    const args = Array.prototype.slice.call(arguments);
    let listener;
    if (typeof args[args.length - 1] === "function") listener = args.pop();

    let port = 0;
    let host = "localhost";
    const first = args[0];
    if (first && typeof first === "object") {
      port = first.port;
      host = first.host || host;
    } else {
      port = first;
      if (typeof args[1] === "string") host = args[1];
    }

    // Go mints and returns the connection id so it never collides with a
    // server-accepted or tls connection.
    const connId = __bento_net_connect(port | 0, host);
    const socket = new Socket(connId);
    socket._connecting = true;
    sockets[connId] = socket;
    if (listener) socket.once("connect", listener);
    return socket;
  }

  class Server extends EventEmitter {
    constructor(options, connectionListener) {
      super();
      if (typeof options === "function") {
        connectionListener = options;
        options = {};
      }
      this._id = __bento_net_createServer();
      servers[this._id] = this;
      this.listening = false;
      this._address = null;
      if (connectionListener) this.on("connection", connectionListener);
    }
    listen() {
      const args = Array.prototype.slice.call(arguments);
      let cb;
      if (typeof args[args.length - 1] === "function") cb = args.pop();
      let port = 0;
      let host = "";
      if (args.length && typeof args[0] === "object" && args[0] !== null) {
        port = args[0].port | 0;
        host = args[0].host || "";
      } else {
        if (args.length >= 1 && args[0] !== undefined) port = args[0] | 0;
        if (args.length >= 2 && typeof args[1] === "string") host = args[1];
      }
      if (cb) this.once("listening", cb);
      __bento_net_listen(this._id, port, host);
      return this;
    }
    address() {
      return this._address;
    }
    close(cb) {
      if (cb) this.once("close", cb);
      __bento_net_closeServer(this._id);
      return this;
    }
  }

  function createServer(options, connectionListener) {
    return new Server(options, connectionListener);
  }

  globalThis.__bento_net_dispatchListening = function (serverId, port, address) {
    const server = servers[serverId];
    if (!server) return;
    server.listening = true;
    server._address = { address: address, port: port, family: address.indexOf(":") >= 0 ? "IPv6" : "IPv4" };
    server.emit("listening");
  };

  globalThis.__bento_net_dispatchConnection = function (serverId, connId, infoJSON) {
    const server = servers[serverId];
    if (!server) return;
    const socket = new Socket(connId);
    socket._applyInfo(JSON.parse(infoJSON));
    sockets[connId] = socket;
    server.emit("connection", socket);
  };

  globalThis.__bento_net_dispatchConnect = function (connId, infoJSON) {
    const socket = sockets[connId];
    if (!socket) return;
    socket._connecting = false;
    socket._applyInfo(JSON.parse(infoJSON));
    socket.emit("connect");
  };

  globalThis.__bento_net_dispatchData = function (connId, b64) {
    const socket = sockets[connId];
    if (!socket) return;
    const buf = Buffer.from(b64, "base64");
    socket.bytesRead += buf.length;
    socket.push(buf);
  };

  globalThis.__bento_net_dispatchEnd = function (connId) {
    const socket = sockets[connId];
    if (!socket) return;
    socket.push(null);
  };

  globalThis.__bento_net_dispatchClose = function (connId) {
    const socket = sockets[connId];
    if (!socket) return;
    delete sockets[connId];
    socket.emit("close");
  };

  globalThis.__bento_net_dispatchError = function (connId, message) {
    const socket = sockets[connId];
    if (!socket) return;
    delete sockets[connId];
    socket.emit("error", new Error(message));
  };

  globalThis.__bento_net_dispatchServerError = function (serverId, code, message) {
    const server = servers[serverId];
    if (!server) return;
    const err = new Error(message);
    if (code) err.code = code;
    server.emit("error", err);
  };

  globalThis.__bento_net_dispatchServerClose = function (serverId) {
    const server = servers[serverId];
    if (!server) return;
    server.listening = false;
    delete servers[serverId];
    server.emit("close");
  };

  module.exports = {
    Server: Server,
    Socket: Socket,
    createServer: createServer,
    connect: connect,
    createConnection: connect,
  };
});
